File: sum_of_squares/sum_of_squares.py
#!/usr/bin/env python3

# https://www.janestreet.com/puzzles/sum-of-squares-index/

import typing

GRID_SIZE = 5

def is_even(num: int) -> bool:
    return num % 2 == 0

def is_odd(num: int) -> bool:
    return num % 2 != 0

def row_to_num(grid: typing.List[int], row: int) -> int:
    return int(''.join(str(n) for n in grid[row * GRID_SIZE : (row + 1) * GRID_SIZE]))

def num_to_row(grid: typing.List[int], row: int, num: int) -> None:
    num_str = str(num).zfill(GRID_SIZE)

    for i in range(GRID_SIZE):
        grid[row * GRID_SIZE + i] = int(num_str[i])

def row_divisible_by(grid: typing.List[int], row: int, number: int) -> bool:
    return row_to_num(grid, row) % number == 0

def col_to_num(grid: typing.List[int], col: int) -> int:
    num_str = ''

    for i in range(GRID_SIZE):
        num_str += str(grid[i * GRID_SIZE + col])
    
    return int(num_str)

def num_to_col(grid: typing.List[int], col: int, num: int) -> None:
    num_str = str(num).zfill(GRID_SIZE)

    for i in range(GRID_SIZE):
        grid[i * GRID_SIZE + col] = int(num_str[i])

def col_divisible_by(grid: typing.List[int], col: int, number: int) -> bool:
    return col_to_num(grid, col) % number == 0

def sum_grid(grid: typing.List[int]) -> int:
    return sum(grid)

row_divisors = (1, 2, 3, 4, 5)
col_divisors = (6, 7, 8, 9, 10)

def valid_grid(grid: typing.List[int]) -> bool:
    for i, row_div in enumerate(row_divisors):
        if i == GRID_SIZE:
            break

        if not row_divisible_by(grid, i, row_div):
            return False

    for i, col_div in enumerate(col_divisors):
        if i == GRID_SIZE:
            break

        if not col_divisible_by(grid, i, col_div):
            return False

    return True

result = None
max_sum = 0
search = 0

def solve_grid(grid: typing.List[int], index: int) -> None:
    global result
    global max_sum
    global search
    search += 1

    if index == (GRID_SIZE * GRID_SIZE):
        if not valid_grid(grid):
            return
        
        s = sum_grid(grid)

        if s > max_sum:
            result = grid.copy()
            max_sum = s

        return

    col = index % GRID_SIZE
    row = index // GRID_SIZE

    for digit in range(9, -1, -1):
        old = grid[index]
        grid[index] = digit

        # prune early searches
        if col == (GRID_SIZE - 1):
            divisor = row_divisors[row]

            if not row_divisible_by(grid, row, divisor):
                continue

        if row == (GRID_SIZE - 1):
            divisor = col_divisors[col]

            if not col_divisible_by(grid, col, divisor):
                continue

        solve_grid(grid, index + 1)

        grid[index] = old

def sum_of_squares() -> None:
    global search

    grid = list(0 for _ in range(GRID_SIZE * GRID_SIZE))

    solve_grid(grid, 0) 

    print(f"RESULT ({search} searches)")
    print(result, max_sum)

if __name__ == "__main__":
    sum_of_squares()

File: sum_of_squares/sum_of_squares.c
/*
    https://www.janestreet.com/puzzles/sum-of-squares-index/
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define GRID_SIZE 5
#define TOTAL_SIZE (GRID_SIZE * GRID_SIZE)

int result[TOTAL_SIZE];
int max_sum = 0;
long search_count = 0;
int row_divisors[] = {1, 2, 3, 4, 5};
int col_divisors[] = {6, 7, 8, 9, 10};
int pow_of_10[] = { 1, 10, 100, 1000, 10000 };

typedef struct {
    int row;
    int col;
} CellPos;

CellPos cell_order[] = {
    {0, 4}, {1, 4}, {2, 4}, {3, 4}, {4, 4},
    {4, 0}, {4, 1}, {4, 2}, {4, 3},
    {0, 0}, {0, 1}, {0, 2}, {0, 3},
    {1, 0}, {1, 1}, {1, 2}, {1, 3},
    {2, 0}, {2, 1}, {2, 2}, {2, 3},
    {3, 0}, {3, 1}, {3, 2}, {3, 3}
};

int cell_order_size = sizeof(cell_order) / sizeof(cell_order[0]);

inline bool is_even(int num) {
    return num % 2 == 0;
}

inline int row_to_num(int grid[], int row) {
    int num = 0;

    for (int i = 0; i < GRID_SIZE; i++) {
        num += grid[row * GRID_SIZE + i] * pow_of_10[4 - i];
    }

    return num;
}

inline bool row_divisible_by(int grid[], int row, int number) {
    return row_to_num(grid, row) % number == 0;
}

inline int col_to_num(int grid[], int col) {
    int num = 0;

    for (int i = 0; i < GRID_SIZE; i++) {
        num += grid[i * GRID_SIZE + col] * pow_of_10[4 - i];
    }

    return num;
}

inline bool col_divisible_by(int grid[], int col, int number) {
    return col_to_num(grid, col) % number == 0;
}

int sum_grid(int grid[]) {
    int sum = 0;

    for (int i = 0; i < TOTAL_SIZE; i++) {
        sum += grid[i];
    }

    return sum;
}

void print_grid(int grid[]) {
    printf("[");

    for (int i = 0; i < GRID_SIZE; i++) {
        for (int j = 0; j < GRID_SIZE; j++)
        {
            printf("%d", grid[i * GRID_SIZE + j]);

            if (j < GRID_SIZE - 1) {
                printf(", ");
            }
        }

        if (i < GRID_SIZE - 1) {
            printf("\n");
        }
    }

    printf("]\n");
}

void solve_grid(int grid[], int index) {
    int total = sum_grid(grid);

    if ((total + (TOTAL_SIZE - index) * 9) <= max_sum)
    {
        return;
    }
    
    if (index == TOTAL_SIZE) {
        bool valid = true;

        for (int r = 0; r < GRID_SIZE && valid; r++) {
            if (!row_divisible_by(grid, r, row_divisors[r])) {
                valid = false;
            }
        }
        
        for (int c = 0; c < GRID_SIZE && valid; c++) {
            if (!col_divisible_by(grid, c, col_divisors[c])) {
                valid = false;
            }
        }
        
        if (valid) {
            if (total > max_sum) {
                max_sum = total;
                memcpy(result, grid, sizeof(int) * TOTAL_SIZE);
            }
        }

        return;
    }

    search_count++;
    
    int row = cell_order[index].row;
    int col = cell_order[index].col;
    int pos = row * GRID_SIZE + col;
    
    for (int digit = 9; digit >= 0; digit--) {
        grid[pos] = digit;
        
        if (col == GRID_SIZE - 1) {
            if (!row_divisible_by(grid, row, row_divisors[row])) {
                continue;
            }
        }
        
        if (row == GRID_SIZE - 1) {
            if (!col_divisible_by(grid, col, col_divisors[col])) {
                continue;
            }
        }
        
        solve_grid(grid, index + 1);
    }
    
    grid[pos] = 0;
}

void sum_of_squares() {
    int grid[TOTAL_SIZE];
    
    memset(grid, 0, TOTAL_SIZE * sizeof(int));
    memset(result, 0, TOTAL_SIZE * sizeof(int));
    
    solve_grid(grid, 0);
    
    printf("RESULT (%ld searches)\n", search_count);

    print_grid(result);
    
    printf("Max sum: %d\n", max_sum);
}

int main() {
    sum_of_squares();
    return 0;
}